relay-server/src/actors/project_cache.rs

```diff
@@ -0,0 +1,250 @@
+use std::collections::{hash_map::Entry, HashMap};
+use std::sync::Arc;
+use std::time::Instant;
+
+use actix::prelude::*;
+use actix_web::ResponseError;
+use failure::Fail;
+use futures::{future, Future};
+use serde::{Deserialize, Serialize};
+
+use relay_common::{metric, ProjectId};
+use relay_config::Config;
+
+use crate::actors::local_project_source::LocalProjectSource;
+use crate::actors::project::{Project, ProjectState};
+use crate::actors::upstream::UpstreamRelay;
+use crate::actors::upstream_project_source::UpstreamProjectSource;
+use crate::metrics::{RelayCounters, RelayHistograms, RelayTimers};
+use crate::utils::{RedisPool, Response};
+
+#[cfg(feature = "processing")]
+use {crate::actors::redis_project_source::RedisProjectSource, relay_common::clone};
+
+#[derive(Fail, Debug)]
+pub enum ProjectError {
+    #[fail(display = "failed to fetch project state from upstream")]
+    FetchFailed,
+
+    #[fail(display = "could not schedule project fetching")]
+    ScheduleFailed(#[cause] MailboxError),
+}
+
+impl ResponseError for ProjectError {}
+
+struct ProjectEntry {
+    last_updated_at: Instant,
+    project: Addr<Project>,
+}
+
+pub struct ProjectCache {
+    config: Arc<Config>,
+    projects: HashMap<ProjectId, ProjectEntry>,
+
+    local_source: Addr<LocalProjectSource>,
+    upstream_source: Addr<UpstreamProjectSource>,
+    #[cfg(feature = "processing")]
+    redis_source: Option<Addr<RedisProjectSource>>,
+}
+
+impl ProjectCache {
+    pub fn new(
+        config: Arc<Config>,
+        upstream_relay: Addr<UpstreamRelay>,
+        _redis: Option<RedisPool>,
+    ) -> Self {
+        let local_source = LocalProjectSource::new(config.clone()).start();
```

Comment:
Please spawn this into its own thread, similar to the `Outcomes` actor. Since it runs quite expensive futures internally, it's better to be isolated.

Reply:
In the old implementation this was also spawned into a thread and then sent a message back to ProjectCache.

Reply:
Yes, the thread is still spawned explicitly in LocalProjectSource. LocalProjectSource's message handlers are not expensive.